unliftio/src/UnliftIO/IO/File/Posix.hs

```diff
@@ -580,3 +582,13 @@ withBinaryFileAtomic filePath iomode action =
       liftIO $ atomicTempFileRename Nothing mFileMode eTmpFile filePath
       pure res
 
+
+-- | See `renameFileDurable`
+renameFileDurable ::
+     MonadIO m => FilePath -> FilePath -> m ()
+renameFileDurable oldName newName =
+     liftIO $ withDirectory (takeDirectory oldName) $ \oldDirFd ->
+                withDirectory (takeDirectory newName) $ \newDirFd -> do
+                  renameFile oldName newName
+                  fsyncDirectoryFd "renameFileDurable" newDirFd
+                  fsyncDirectoryFd "renameFileDurable" oldDirFd
```

Comment:
Another possible improvement could be to use `renameat()`, with is directory-FD based renaming.

That would be even better than what we have here, because then the `fsync()` would operate on the same FD as the rename does, giving even better guarantees (see also https://stackoverflow.com/questions/37288453/calling-fsync2-after-close2/50158433#50158433 for a related topic on files).

That is not necessary for this PR though, having a `renameFileDurable` implemented this way is already much better than not having it.